File: apps/api/src/modules/kpi/kpi-import.service.ts
import { Injectable, BadRequestException } from '@nestjs/common';
import {
  getMssqlDb,
  kpiAgreementAreas,
  kpiAgreementMetrics,
  eq,
} from '@lotris/db';
import { v4 as uuidv4 } from 'uuid';
import * as ExcelJS from 'exceljs';
import type { TrpcAuth } from '@lotris/types';
import type { ImportColumnMappingDto } from './dto';
import { KpiService } from './kpi.service';

interface ParsedRow {
  description: string;
  weight: number;
  targetScore: number;
  measurementPeriod: string;
  kpiDefinitionId?: string;
}

@Injectable()
export class KpiImportService {
  // In-memory cache of parsed rows per agreement (keyed by agreementId)
  // Real production use would store in Redis/temp table; this is adequate for the sprint
  private pendingRows = new Map<string, ParsedRow[]>();

  constructor(private readonly kpi: KpiService) {}

  /**
   * Parse an uploaded Excel/CSV file and return a column preview.
   * File is cached in memory for the subsequent import call.
   */
  async parseUpload(auth: TrpcAuth, agreementId: string, file: Express.Multer.File) {
    await this.kpi.getAgreement(auth, agreementId); // access check

    if (!file) throw new BadRequestException('No file uploaded');
    const ext = file.originalname.split('.').pop()?.toLowerCase();
    if (!['xlsx', 'xls', 'csv'].includes(ext ?? '')) {
      throw new BadRequestException('Only .xlsx, .xls, or .csv files are accepted');
    }

    const workbook = new ExcelJS.Workbook();
    if (ext === 'csv') {
      // ExcelJS CSV reader accepts a readable stream; convert buffer via Readable.from
      // eslint-disable-next-line @typescript-eslint/no-var-requires
      const { Readable } = require('node:stream') as typeof import('stream');
      await workbook.csv.read(Readable.from(file.buffer));
    } else {
      await workbook.xlsx.load(Buffer.from(file.buffer));
    }

    const sheet = workbook.worksheets[0];
    if (!sheet) throw new BadRequestException('Spreadsheet has no worksheets');

    // Return first row as column headers + up to 5 sample data rows
    const headers: string[] = [];
    const sampleRows: Record<string, string>[] = [];

    sheet.eachRow((row, rowIndex) => {
      if (rowIndex === 1) {
        row.eachCell((cell) => headers.push(String(cell.value ?? '')));
        return;
      }
      if (rowIndex <= 6) {
        const rowData: Record<string, string> = {};
        row.eachCell((cell, colIndex) => {
          const header = headers[colIndex - 1] ?? `col${colIndex}`;
          rowData[header] = String(cell.value ?? '');
        });
        sampleRows.push(rowData);
      }
    });

    // Store raw parsed data for later import
    const allRows: ParsedRow[] = [];
    sheet.eachRow((row, rowIndex) => {
      if (rowIndex === 1) return;
      const cells: string[] = [];
      row.eachCell((cell) => cells.push(String(cell.value ?? '')));
      allRows.push({
        description: cells[0] ?? '',
        weight: Number(cells[1] ?? 0),
        targetScore: Number(cells[2] ?? 0),
        measurementPeriod: 'MONTHLY',
      });
    });
    this.pendingRows.set(agreementId, allRows);

    return { headers, sampleRows, totalRows: allRows.length };
  }

  /**
   * Apply column mapping and insert metric rows into the first area of the agreement.
   * Creates a default area "Imported Metrics" if none exists.
   */
  async importWithMapping(auth: TrpcAuth, agreementId: string, _dto: ImportColumnMappingDto) {
    const agreement = await this.kpi.getAgreement(auth, agreementId);
    if (agreement.status !== 'DRAFT') {
      throw new BadRequestException('Can only import into DRAFT agreements');
    }

    const rows = this.pendingRows.get(agreementId);
    if (!rows || rows.length === 0) {
      throw new BadRequestException('No pending upload found. Upload a file first.');
    }

    const db = await getMssqlDb();
    // Get or create a default import area
    const [existingArea] = await db
      .select()
      .from(kpiAgreementAreas)
      .where(eq(kpiAgreementAreas.agreementId, agreementId));

    let areaId: string;
    if (existingArea) {
      areaId = existingArea.id;
    } else {
      areaId = uuidv4();
      await db.insert(kpiAgreementAreas).values({
        id: areaId,
        tenantId: agreement.tenantId,
        agreementId,
        name: 'Imported Metrics',
        weight: '100',
        sortOrder: 0,
      });
    }

    // Insert metric rows
    for (let i = 0; i < rows.length; i++) {
      const r = rows[i]!;
      if (!r.description || r.targetScore <= 0) continue;
      await db.insert(kpiAgreementMetrics).values({
        id: uuidv4(),
        tenantId: agreement.tenantId,
        areaId,
        kpiDefinitionId: r.kpiDefinitionId ?? null,
        description: r.description,
        measurementPeriod: r.measurementPeriod ?? 'MONTHLY',
        weight: String(r.weight),
        targetScore: String(r.targetScore),
        actualScore: null,
        sortOrder: i,
      });
    }

    this.pendingRows.delete(agreementId);
    return { imported: rows.length };
  }
}

File: src/Lotris.Application/Kpi/KpiImportService.cs
using Lotris.Application.Common;
using Lotris.Contracts;
using Lotris.Contracts.Kpi;
using System.Globalization;

namespace Lotris.Application.Kpi;

public sealed class KpiImportService
{
    private readonly KpiService _kpi;
    private readonly IKpiRepository _repository;
    private readonly IKpiImportPendingStore _pending;
    private readonly IKpiSpreadsheetParser _parser;

    public KpiImportService(
        KpiService kpi,
        IKpiRepository repository,
        IKpiImportPendingStore pending,
        IKpiSpreadsheetParser parser)
    {
        _kpi = kpi;
        _repository = repository;
        _pending = pending;
        _parser = parser;
    }

    public async Task<KpiUploadPreviewResponse> ParseUploadAsync(
        LotrisSession session,
        Guid agreementId,
        Stream fileStream,
        string fileName,
        CancellationToken cancellationToken = default)
    {
        await _kpi.RequireAgreementForImportAsync(session, agreementId, cancellationToken);

        var ext = Path.GetExtension(fileName).TrimStart('.').ToLowerInvariant();
        if (ext is not ("xlsx" or "xls" or "csv"))
        {
            throw new BadRequestException("Only .xlsx, .xls, or .csv files are accepted");
        }

        var table = _parser.Parse(fileStream, fileName);

        if (table.Headers.Count == 0)
        {
            throw new BadRequestException("Spreadsheet has no worksheets or headers");
        }

        var allRows = new List<KpiPendingImportRow>();
        for (var i = 0; i < table.Rows.Count; i++)
        {
            var cells = table.Rows[i].ToList();
            allRows.Add(new KpiPendingImportRow
            {
                Description = GetCell(cells, 0),
                Weight = ParseDecimal(GetCell(cells, 1)),
                TargetScore = ParseDecimal(GetCell(cells, 2)),
                MeasurementPeriod = "MONTHLY",
            });
        }

        _pending.Set(agreementId, allRows);

        var sampleRows = table.Rows.Take(5).Select(row =>
        {
            var dict = new Dictionary<string, string>();
            for (var i = 0; i < table.Headers.Count; i++)
            {
                var header = table.Headers[i];
                dict[header] = i < row.Count ? row[i] : string.Empty;
            }

            return dict;
        }).ToList();

        return new KpiUploadPreviewResponse(table.Headers, sampleRows, allRows.Count);
    }

    public async Task<KpiImportResultResponse> ImportWithMappingAsync(
        LotrisSession session,
        Guid agreementId,
        ImportColumnMappingRequest request,
        CancellationToken cancellationToken = default)
    {
        var agreement = await _kpi.RequireAgreementForImportAsync(session, agreementId, cancellationToken);
        if (agreement.Status != "DRAFT")
        {
            throw new BadRequestException("Can only import into DRAFT agreements");
        }

        var rows = _pending.Get(agreementId);
        if (rows is null || rows.Count == 0)
        {
            throw new BadRequestException("No pending upload found. Upload a file first.");
        }

        var mappedRows = rows.Select(row => new KpiPendingImportRow
        {
            Description = row.Description,
            Weight = row.Weight,
            TargetScore = row.TargetScore,
            MeasurementPeriod = row.MeasurementPeriod,
            KpiDefinitionId = row.KpiDefinitionId,
        }).ToList();

        var existingArea = await _repository.GetFirstAgreementAreaAsync(agreementId, cancellationToken);
        Guid areaId;
        if (existingArea is not null)
        {
            areaId = existingArea.Id;
        }
        else
        {
            areaId = Guid.NewGuid();
            await _repository.CreateAgreementAreaAsync(new KpiAgreementAreaEntity
            {
                Id = areaId,
                TenantId = agreement.TenantId,
                AgreementId = agreementId,
                Name = "Imported Metrics",
                Weight = 100m,
                SortOrder = 0,
            }, cancellationToken);
        }

        var imported = 0;
        for (var i = 0; i < mappedRows.Count; i++)
        {
            var row = mappedRows[i];
            if (string.IsNullOrWhiteSpace(row.Description) || row.TargetScore <= 0)
            {
                continue;
            }

            await _repository.CreateAgreementMetricAsync(new KpiAgreementMetricEntity
            {
                Id = Guid.NewGuid(),
                TenantId = agreement.TenantId,
                AreaId = areaId,
                KpiDefinitionId = row.KpiDefinitionId,
                Description = row.Description,
                MeasurementPeriod = row.MeasurementPeriod,
                Weight = row.Weight,
                TargetScore = row.TargetScore,
                ActualScore = null,
                SortOrder = i,
            }, cancellationToken);
            imported++;
        }

        _pending.Remove(agreementId);
        return new KpiImportResultResponse(imported);
    }

    private static string GetCell(IReadOnlyList<string> cells, int index) =>
        index < cells.Count ? cells[index] : string.Empty;

    private static decimal ParseDecimal(string value) =>
        decimal.TryParse(value, NumberStyles.Any, CultureInfo.InvariantCulture, out var parsed) ? parsed : 0m;
}
